### PR TITLE
 api(ticdc): add /api/v2/tso: get tso from pd 

### LINKED ISSUE #5586: TiCDC Open API V2: return tso get from pd
provide an api which returns the pd tso
response like:
```go
type Tso struct {
   Timestamp int64 `json:"timestamp"`
   LogicTime int64 `json:"logic-time"`
}
```

### PR DESCRIPTION
**Pull request by crelax**
TiCDC Open API V2: return tso get from pd 

### What problem does this PR solve?

Issue Number: close #5586

### What is changed and how it works?
Get tso from PD 

#### Tests 

 - Unit test
 - Integration test

##### Will it cause performance regression or break compatibility?
No
##### Do you need to update user documentation, design documentation or monitoring documentation?
No
### Release note 

```release-note
TiCDC Open API V2: return tso get from pd 
```

### CODE REVIEW

File: cdc/api/v2/api.go [hunk 1]
```diff
@@ -14,25 +14,62 @@
...
+	pdClient := h.capture.UpstreamManager.Get(upstream.DefaultUpstreamID).PDClient
+	timestamp, logictTime, err := pdClient.GetTS(ctx)
+	if err != nil {
+		_ = c.Error(err)
+		c.IndentedJSON(http.StatusInternalServerError, model.NewHTTPError(err))
+		return
+	}
+
+	resp := Tso{timestamp, logictTime}
+	//resp := Tso{0, 0}
```
Comment (liuzix) [inline cdc/api/v2/api.go:73]:
```suggestion
```

File: cdc/api/v2/api.go [hunk 2]
```diff
@@ -14,25 +14,62 @@
...
 	"github.com/pingcap/tiflow/cdc/capture"
+	"github.com/pingcap/tiflow/cdc/owner"
 )
 
 // OpenAPIV2 provides CDC v2 APIs
-type OpenAPIV2 struct {
+type openAPIV2 struct {
 	capture *capture.Capture
+	// use for unit test only
+	testStatusProvider owner.StatusProvider
```
Comment (liuzix) [inline cdc/api/v2/api.go:31]: I think adding a field solely for the purpose of testing is not good for code maintenance. Can you consider using a `StatusProvider` in production code path too? Maybe we can remove the dependence on `capture.Capture`.

File: cdc/api/v2/api.go [hunk 3]
```diff
@@ -14,25 +14,62 @@
...
 	v2.Use(middleware.ErrorHandleMiddleware())
+
+	// common APIs
+	v2.GET("/tso", api.GetTso)
+}
+
+func (h *openAPIV2) GetTso(c *gin.Context) {
+	ctx := c.Request.Context()
+	pdClient := h.capture.UpstreamManager.Get(upstream.DefaultUpstreamID).PDClient
+	timestamp, logictTime, err := pdClient.GetTS(ctx)
```
Comment (liuzix) [inline cdc/api/v2/api.go:65]:
```suggestion
	timestamp, logicalTime, err := pdClient.GetTS(ctx)
```

File: cdc/api/v2/api.go [hunk 4]
```diff
@@ -14,25 +14,62 @@
 package v2
 
 import (
+	"github.com/pingcap/tiflow/cdc/model"
+	"github.com/pingcap/tiflow/pkg/upstream"
+	"net/http"
```
Comment (liuzix) [inline cdc/api/v2/api.go:19]: Please group imports

File: cdc/api/v2/api.go [hunk 5]
```diff
@@ -14,25 +14,62 @@
...
+
 	"github.com/gin-gonic/gin"
 	"github.com/pingcap/tiflow/cdc/api/middleware"
 	"github.com/pingcap/tiflow/cdc/capture"
+	"github.com/pingcap/tiflow/cdc/owner"
 )
 
 // OpenAPIV2 provides CDC v2 APIs
-type OpenAPIV2 struct {
+type openAPIV2 struct {
```
Comment (liuzix) [inline cdc/api/v2/api.go:28]: Why is this struct private?

File: cdc/api/v2/api.go [hunk 6]
```diff
@@ -14,25 +14,62 @@
...
 	capture *capture.Capture
+	// use for unit test only
+	testStatusProvider owner.StatusProvider
+}
+
+// NewOpenAPIV2 creates a new openAPIV2.
+func NewOpenAPIV2(c *capture.Capture) openAPIV2 {
```
Comment (liuzix) [inline cdc/api/v2/api.go:35]: Public function returning a private type?

File: cdc/api/v2/api.go (hunk 5 above)
Comment (crelax) [reply cdc/api/v2/api.go:28]: hmm, it seems that being public makes more sense. I edited it private since it's private in API v1.

**Comment by sdojjy** [comment]
please adjust your IDE code format style to fix the lint.

File: cdc/api/v2/api.go [hunk 7]
```diff
@@ -14,25 +14,61 @@
...
 	v2.Use(middleware.LogMiddleware())
 	v2.Use(middleware.ErrorHandleMiddleware())
+
+	// common APIs
+	v2.GET("/tso", api.GetTso)
+}
+
+func (h *openAPIV2) GetTso(c *gin.Context) {
+	ctx := c.Request.Context()
+	pdClient := h.capture.UpstreamManager.Get(upstream.DefaultUpstreamID).PDClient
```
Comment (liuzix) [inline cdc/api/v2/api.go:64]: Can you check whether `UpstreamManager` or `UpstreamManager.Get(upstream.DefaultUpstreamID)` or `PDClient` is `nil`?

File: cdc/api/v2/api_test.go [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+
+func TestGetTso(t *testing.T) {
+	t.Parallel()
+	mockPDClient := &MockPDClient{}
+	mockManager := upstream.NewManager4Test(mockPDClient)
+	cp := capture.NewCaptureWithManager4Test(mockManager)
+	router := newRouter(cp, newStatusProvider())
+	w := httptest.NewRecorder()
+
+	api := testCase{url: "/api/v2/tso", method: "GET"}
```
Comment (liuzix) [inline cdc/api/v2/api_test.go:113]: Why is a variable of type `testCase` is named `api`?

File: cdc/api/v2/api.go (hunk 2 above)
Comment (crelax) [reply cdc/api/v2/api.go:31]: will remove the dependence on *capture later

File: cdc/api/v2/api.go (hunk 7 above)
Comment (crelax) [reply cdc/api/v2/api.go:69]: ok

File: cdc/api/v2/api_test.go (hunk 8 above)
Comment (crelax) [reply cdc/api/v2/api_test.go]: ok

**Comment by liuzix** [comment]
Is there an integration (e2e) test case for this feature?

File: tests/integration_tests/http_api/util/test_case.py [hunk 9]
```diff
@@ -298,6 +300,19 @@ def set_log_level():
 
     print("pass test: set log level")
 
+def get_tso():
+    # test state: all
+    url = V2_BASE_URL0+"/tso"
+
+    for i in range(RETRY_TIME):
```
Comment (liuzix) [inline tests/integration_tests/http_api/util/test_case.py:307]: What is the motivation for retrying here? What can happen if we don't retry?

Comment (crelax) [reply tests/integration_tests/http_api/util/test_case.py:307]: loop deleted

**Review by liuzix — APPROVED** [review]

**Pull request merged**